File: 3.x.x/crates-utopia/dns/src/validator/caa.rs
use parking_lot::Mutex;
use serde_json::Value;
use utopia_validators::{Validator, ValueType};

/// PHP `Utopia\DNS\Validator\CAA`.
#[derive(Debug)]
pub struct CAA {
    reason: Mutex<String>,
}

impl Default for CAA {
    fn default() -> Self {
        Self::new()
    }
}

impl CAA {
    pub const CAA_FLAG_MIN: i32 = 0;
    pub const CAA_FLAG_MAX: i32 = 255;
    pub const FAILURE_REASON_INVALID_FLAGS: &'static str =
        "Flags must be a number between 0 and 255";
    pub const FAILURE_REASON_INVALID_TAG: &'static str = "Tag must be a non-empty string";
    pub const FAILURE_REASON_INVALID_VALUE: &'static str =
        "Value must be a non-empty string and must be enclosed in quotes";
    pub const FAILURE_REASON_INVALID_FORMAT: &'static str =
        "CAA record must be in the format <flags> <tag> \"<value>\"";

    #[must_use]
    pub fn new() -> Self {
        Self {
            reason: Mutex::new(String::new()),
        }
    }

    pub fn reason(&self) -> String {
        self.reason.lock().clone()
    }
}

impl Validator for CAA {
    fn description(&self) -> String {
        let reason = self.reason.lock();
        if !reason.is_empty() && *reason != "0" {
            reason.clone()
        } else {
            Self::FAILURE_REASON_INVALID_FORMAT.to_string()
        }
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, data: &Value) -> bool {
        let Some(data) = data.as_str() else {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_FORMAT.into();
            return false;
        };
        let mut parts = data.splitn(3, ' ');
        let (Some(flags), Some(tag), Some(value)) = (parts.next(), parts.next(), parts.next())
        else {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_FORMAT.into();
            return false;
        };
        if !is_numeric(flags) {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_FLAGS.into();
            return false;
        }
        let flags_n: i32 = flags.parse().unwrap_or(i32::MIN);
        if !(Self::CAA_FLAG_MIN..=Self::CAA_FLAG_MAX).contains(&flags_n) {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_FLAGS.into();
            return false;
        }
        if tag.is_empty() {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_TAG.into();
            return false;
        }
        if value.is_empty() || !value.starts_with('"') || !value.ends_with('"') {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_VALUE.into();
            return false;
        }
        let inner = &value[1..value.len() - 1];
        if inner.is_empty() {
            *self.reason.lock() = Self::FAILURE_REASON_INVALID_VALUE.into();
            return false;
        }
        true
    }
}

fn is_numeric(s: &str) -> bool {
    !s.is_empty() && s.parse::<f64>().is_ok()
}

File: packages/dns/src/Validator/CAA.php
<?php

declare(strict_types=1);

namespace Utopia\DNS\Validator;

use Utopia\Validator;

class CAA extends Validator
{
    public const int CAA_FLAG_MIN = 0;

    public const int CAA_FLAG_MAX = 255;

    public const string FAILURE_REASON_INVALID_FLAGS = 'Flags must be a number between 0 and 255';

    public const string FAILURE_REASON_INVALID_TAG = 'Tag must be a non-empty string';

    public const string FAILURE_REASON_INVALID_VALUE = 'Value must be a non-empty string and must be enclosed in quotes';

    public const string FAILURE_REASON_INVALID_FORMAT = 'CAA record must be in the format <flags> <tag> "<value>"';

    public string $reason = '';

    /**
     * Check if the provided value matches the CAA record format
     */
    public function isValid(mixed $data): bool
    {
        if (!\is_string($data)) {
            $this->reason = self::FAILURE_REASON_INVALID_FORMAT;
            return false;
        }

        $parts = explode(' ', $data, 3);

        if (\count($parts) !== 3) {
            $this->reason = self::FAILURE_REASON_INVALID_FORMAT;
            return false;
        }

        $flags = $parts[0];
        $tag = $parts[1];
        $value = $parts[2];

        // Check flags is a number
        if (!is_numeric($flags)) {
            $this->reason = self::FAILURE_REASON_INVALID_FLAGS;
            return false;
        }

        $flags = (int) $flags;

        // Check flags is within the allowed range
        if ($flags < self::CAA_FLAG_MIN || $flags > self::CAA_FLAG_MAX) {
            $this->reason = self::FAILURE_REASON_INVALID_FLAGS;
            return false;
        }

        // Check tag is not empty
        if ($tag === '') {
            $this->reason = self::FAILURE_REASON_INVALID_TAG;
            return false;
        }

        // Check value is not empty and starts with " and ends with "
        if ($value === '' || $value[0] !== '"' || $value[\strlen($value) - 1] !== '"') {
            $this->reason = self::FAILURE_REASON_INVALID_VALUE;
            return false;
        }

        $value = substr($value, 1, \strlen($value) - 2);

        // Check value is not empty after removing the quotes
        if ($value === '') {
            $this->reason = self::FAILURE_REASON_INVALID_VALUE;
            return false;
        }

        // All checks passed
        return true;
    }

    public function getDescription(): string
    {
        if ($this->reason !== '' && $this->reason !== '0') {
            return $this->reason;
        }

        return self::FAILURE_REASON_INVALID_FORMAT;
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }
}
